libp2p/connmanager.nim

```diff
@@ -0,0 +1,276 @@
+           maxConnsPerPeer: int = MaxConnectionsPerPeer): ConnManager =
+  C(maxConns: maxConnsPerPeer,
+    conns: initTable[PeerID, HashSet[Connection]](),
+    muxed: initTable[Connection, MuxerHolder]())
+
+proc contains*(c: ConnManager, conn: Connection): bool =
+  ## checks if a connection is being tracked by the
+  ## connection manager
+  ##
+
+  if isNil(conn):
+    return
+
+  if isNil(conn.peerInfo):
+    return
+
+  if conn.peerInfo.peerId notin c.conns:
+    return
+
+  return conn in c.conns[conn.peerInfo.peerId]
+
+proc contains*(c: ConnManager, peerId: PeerID): bool =
+  peerId in c.conns
+
+proc contains*(c: ConnManager, muxer: Muxer): bool =
+  ## checks if a muxer is being tracked by the connection
+  ## manager
+  ##
+
+  if isNil(muxer):
+    return
+
+  let conn = muxer.connection
+  if conn notin c:
+    return
+
+  if conn notin c.muxed:
+    return
+
+  return muxer == c.muxed[conn].muxer
+
+proc cleanupConn(c: ConnManager, conn: Connection) {.async.} =
+  ## clean connection's resources such as muxers and streams
+  ##
+
+  if isNil(conn):
+    return
+
+  if isNil(conn.peerInfo):
+    return
+
+  let peerInfo = conn.peerInfo
+  let lock = c.cleanUpLock.mgetOrPut(peerInfo, newAsyncLock())
+
+  try:
+    await lock.acquire()
+    trace "cleaning up connection for peer", peer = $peerInfo
+    if conn in c.muxed:
+      let muxerHolder = c.muxed[conn]
+      c.muxed.del(conn)
+
+      await muxerHolder.muxer.close()
+      if not(isNil(muxerHolder.handle)):
+        await muxerHolder.handle
+
+    if peerInfo.peerId in c.conns:
+      c.conns[peerInfo.peerId].excl(conn)
+
+      if c.conns[peerInfo.peerId].len == 0:
+        c.conns.del(peerInfo.peerId)
+
+    if not(conn.peerInfo.isClosed()):
+      conn.peerInfo.close()
+
+  finally:
+    await conn.close()
+    libp2p_peers.set(c.conns.len.int64)
+
+    if lock.locked():
+      lock.release()
+
+    trace "connection cleaned up"
+
+proc onClose(c: ConnManager, conn: Connection) {.async.} =
+  ## connection close even handler
+  ##
+  ## triggers the connections resource cleanup
+  ##
+
+  await conn.closeEvent.wait()
+  trace "triggering connection cleanup"
+  await c.cleanupConn(conn)
+
+proc selectConn*(c: ConnManager,
+                peerInfo: PeerInfo,
+                dir: Direction): Connection =
+  ## Select a connection for the provided peer and direction
+  ##
+
+  if isNil(peerInfo):
+    return
+
+  let conns = toSeq(
+    c.conns.getOrDefault(peerInfo.peerId))
+    .filterIt( it.dir == dir )
+
+  if conns.len > 0:
+    return conns[0]
+
+proc selectConn*(c: ConnManager, peerInfo: PeerInfo): Connection =
+  ## Select a connection for the provided giving priority
+  ## to outgoing connections
+  ##
+
+  if isNil(peerInfo):
+    return
+
+  var conn = c.selectConn(peerInfo, Direction.Out)
+  if isNil(conn):
+    conn = c.selectConn(peerInfo, Direction.In)
+
+  return conn
+
+proc selectMuxer*(c: ConnManager, conn: Connection): Muxer =
```

Comment:
again I would refrain from mixing nim return magic/weirdness

Like here I don't even know what happens, we return Nil? why not use `Result[Muxer,cstring]` since we got no async?

Reply:
We can, I just haven't used results, but the `cstring` there is puzzling, why not a regular string? Also, the muxer itself never know nor care what it's string is, it's only ever relevant when selecting.